// renderkid/lib/renderKid/styles/rule/DeclarationBlock.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

// Generated by CoffeeScript 2.5.1
var Arbitrary, DeclarationBlock, declarationClasses;

module.exports = DeclarationBlock = function () {
  var self;

  var DeclarationBlock = /*#__PURE__*/function () {
    function DeclarationBlock() {
      _classCallCheck(this, DeclarationBlock);

      this._declarations = {};
    }

    _createClass(DeclarationBlock, [{
      key: "set",
      value: function set(prop, value) {
        var key, val;

        if (_typeof(prop) === 'object') {
          for (key in prop) {
            val = prop[key];
            this.set(key, val);
          }

          return this;
        }

        prop = self.sanitizeProp(prop);

        this._getDeclarationClass(prop).setOnto(this._declarations, prop, value);

        return this;
      }
    }, {
      key: "_getDeclarationClass",
      value: function _getDeclarationClass(prop) {
        var cls;

        if (prop[0] === '_') {
          return Arbitrary;
        }

        if (!(cls = declarationClasses[prop])) {
          throw Error("Unknown property `".concat(prop, "`. Write it as `_").concat(prop, "` if you're defining a custom property"));
        }

        return cls;
      }
    }], [{
      key: "sanitizeProp",
      value: function sanitizeProp(prop) {
        return String(prop).trim();
      }
    }]);

    return DeclarationBlock;
  }();

  ;
  self = DeclarationBlock;
  return DeclarationBlock;
}.call(void 0);

Arbitrary = require('renderkid/lib/renderKid/styles/rule/declarationBlock/Arbitrary');
declarationClasses = {
  color: require('renderkid/lib/renderKid/styles/rule/declarationBlock/Color'),
  background: require('renderkid/lib/renderKid/styles/rule/declarationBlock/Background'),
  width: require('renderkid/lib/renderKid/styles/rule/declarationBlock/Width'),
  height: require('renderkid/lib/renderKid/styles/rule/declarationBlock/Height'),
  bullet: require('renderkid/lib/renderKid/styles/rule/declarationBlock/Bullet'),
  display: require('renderkid/lib/renderKid/styles/rule/declarationBlock/Display'),
  margin: require('renderkid/lib/renderKid/styles/rule/declarationBlock/Margin'),
  marginTop: require('renderkid/lib/renderKid/styles/rule/declarationBlock/MarginTop'),
  marginLeft: require('renderkid/lib/renderKid/styles/rule/declarationBlock/MarginLeft'),
  marginRight: require('renderkid/lib/renderKid/styles/rule/declarationBlock/MarginRight'),
  marginBottom: require('renderkid/lib/renderKid/styles/rule/declarationBlock/MarginBottom'),
  padding: require('renderkid/lib/renderKid/styles/rule/declarationBlock/Padding'),
  paddingTop: require('renderkid/lib/renderKid/styles/rule/declarationBlock/PaddingTop'),
  paddingLeft: require('renderkid/lib/renderKid/styles/rule/declarationBlock/PaddingLeft'),
  paddingRight: require('renderkid/lib/renderKid/styles/rule/declarationBlock/PaddingRight'),
  paddingBottom: require('renderkid/lib/renderKid/styles/rule/declarationBlock/PaddingBottom')
};
