// webpack/lib/dependencies/LoaderImportDependency.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const ModuleDependency = require("webpack/lib/dependencies/ModuleDependency");

/** @typedef {import("webpack/lib/Dependency").GetConditionFn} GetConditionFn */
/** @typedef {import("webpack/lib/ModuleGraph")} ModuleGraph */
/** @typedef {import("webpack/lib/ModuleGraphConnection")} ModuleGraphConnection */
/** @typedef {import("webpack/lib/ModuleGraphConnection").ConnectionState} ConnectionState */
/** @typedef {import("webpack/lib/util/runtime").RuntimeSpec} RuntimeSpec */

class LoaderImportDependency extends ModuleDependency {
	/**
	 * @param {string} request request string
	 */
	constructor(request) {
		super(request);
		this.weak = true;
	}

	get type() {
		return "loader import";
	}

	get category() {
		return "loaderImport";
	}

	/**
	 * @param {ModuleGraph} moduleGraph module graph
	 * @returns {null | false | GetConditionFn} function to determine if the connection is active
	 */
	getCondition(moduleGraph) {
		return false;
	}
}

module.exports = LoaderImportDependency;
